datafusion/optimizer/src/push_down_filter.rs

```diff
@@ -668,16 +670,40 @@ impl OptimizerRule for PushDownFilter {
 
                         (field.qualified_name(), expr)
                     })
-                    .collect::<HashMap<_, _>>();
+                    .partition(|(_, value)| matches!(value, Expr::ScalarFunction(f) if f.fun.volatility() == Volatility::Volatile));
 
-                // re-write all filters based on this projection
-                // E.g. in `Filter: b\n  Projection: a > 1 as b`, we can swap them, but the filter must be "a > 1"
-                let new_filter = LogicalPlan::Filter(Filter::try_new(
-                    replace_cols_by_name(filter.predicate.clone(), &replace_map)?,
-                    projection.input.clone(),
-                )?);
+                let mut push_predicates = vec![];
+                let mut keep_predicates = vec![];
+                for expr in split_conjunction_owned(filter.predicate.clone()).into_iter()
+                {
+                    if contain(expr.clone(), &volatile_map.clone())? {
```

Comment:
I don't think we need to clone here:

```suggestion
                    if contain(expr.clone(), volatile_map)? {
```